### PR TITLE
[Feature] Add support for middlewares

### PR DESCRIPTION
**Pull request by samuelnogueira**
(see https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/architecture.html#middlewares)

### CODE REVIEW

File: example/full-config.php
Comment (Ocramius) [inline example/full-config.php]: Needs documentation/example/comment

File: src/ConfigurationFactory.php
Comment (Ocramius) [inline src/ConfigurationFactory.php]: `$middleware` should be assumed to be a `string` here - an `assert()` is preferable to an `is_string()` check

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/ConfigurationFactory.php
Comment (Ocramius) [inline src/ConfigurationFactory.php]: IMO not to be handled - should always be `string` in config, in my opinion :thinking:

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: example/full-config.php
Comment (Ocramius) [inline example/full-config.php:63]: Possibly worth adding a link to the documentation (in `doctrine/dbal`) for middleware :)

**Review by Ocramius — APPROVED** [review]
LGTM - needs @asgrim's review :)

File: src/ConfigurationFactory.php
Comment (samuelnogueira) [reply src/ConfigurationFactory.php]: That wouldn't conform I think with other configuration values that are accepted as both a container id string, or as the actual instance (ex. https://github.com/Roave/psr-container-doctrine/blob/3.5.x/src/ConfigurationFactory.php#L116).

But, maybe that's a "feature" no one needs, I've made the change to accept only container id strings :+1:

Comment (Ocramius) [reply src/ConfigurationFactory.php]: @samuelnogueira yeah, this library has been passed from hand to hand, so I'm trying to keep it slim/simple, heh :D 

The more we add, the more we have to support later on, and the harder it becomes to migrate, so I'm really trying to cut everything that isn't strictly vital.

File: example/full-config.php
Comment (samuelnogueira) [inline example/full-config.php]: Couldn't find a static link to the current latest version (only for version 3.2, https://www.doctrine-project.org/projects/doctrine-dbal/en/3.2/reference/architecture.html#middlewares), is this OK?

Comment (Ocramius) [reply example/full-config.php]: We can link the RST via absolute sha1 reference perhaps? https://github.com/doctrine/dbal/blob/baf6efed70e46f4d979e5dd42f804007e75f12fb/docs/en/reference/architecture.rst#middlewares

Comment (samuelnogueira) [reply example/full-config.php]: Perfect, updated the documentation link with the tag of latest release (3.3.2): https://github.com/doctrine/dbal/blob/3.3.2/docs/en/reference/architecture.rst#middlewares

Comment (Ocramius) [reply example/full-config.php]: Awesome, thanks!

**Comment by Ocramius** [comment]
Ah, I think `ContainerInterface#has()` must be explicitly stubbed for some calls in the new tests. 

EDIT: sorry for this late feedback - I didn't realize that Github Actions didn't run at all :|

**Comment by Ocramius** [comment]
All green, waiting for another review, and then we can merge+release 👍

**Review by asgrim — APPROVED** [review]
LGTM :+1:

**Comment by asgrim** [comment]
Thanks @samuelnogueira and @Ocramius !

**Comment by Ocramius** [comment]
Cutting a release then, thanks @asgrim!

**Pull request merged**
